### PR TITLE
Feature rank

### PR DESCRIPTION
**Pull request by blackbbc**

### CODE REVIEW

File: vj4/cron_jobs/rank.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+CONN = MongoClient(DB_HOST, DB_PORT)
+DB = CONN[DB_NAME]
+COLL = DB[COLL_NAME]
+
+# Key represents level
+# Value represents percent
+# E.g. "10": 1 means that people who rank in 1% will get 10 levels
+level_config = {
+    "10": 1,
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: 10: 1

File: vj4/cron_jobs/rank.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+from pymongo import MongoClient
+
+DB_HOST = 'localhost'
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: In the future this needs to be command line arguments.

File: vj4/cron_jobs/rank.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  global level_config
+
+  perc *= 100
+
+  for key, value in level_config.items():
+    if perc <= value:
+      return int(key)
+
+def handle_rank():
+  udocs = list(COLL.find())
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: This is evil.

File: vj4/cron_jobs/rank.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+def rank_data(udocs):
+  global level_config
+
+  stack = list()
+  result = list()
+  rank = 1
+  for udoc in udocs:
+    if len(stack) == 0:
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: Pythonically this should be "if not stack:"

File: vj4/cron_jobs/rank.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  rank_array = rank_data(udocs)
+  total = rank_array[-1]
+
+  index = 0
+  for udoc in udocs:
+    rankN = rank_array[index]
+    udoc['rankN'] = rankN
+    udoc['level'] = count_level(rankN / total)
+    COLL.save(udoc)
```
Comment (iceboy-sjtu) [inline vj4/cron_jobs/rank.py]: This is more evil, in several ways:

- You loads all users at once, but saves one by one.
- This is a cron job running along with the main server, and you saved a user document which is loaded several minutes ago.

File: vj4/cron_jobs/rank.py (hunk 3 above)
Comment (blackbbc) [inline vj4/cron_jobs/rank.py]: what does evil means?

Comment (twd2) [inline vj4/cron_jobs/rank.py]: 邪恶的

**Pull request closed without merging**